### PR TITLE
crystal: reduce closure size of compiled exes by splitting outputs

### PR DESCRIPTION
**Pull request by manveru**
###### Motivation for this change

So far, compiled crystal programs retained a reference to `crystal`, which in turn depends on `llvm` and a ton of other things.
This makes sure that the output paths of crystal are  split up, so only `lib` is still referenced for runtime backtrace output.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @david50407 @peterhoeg

### CODE REVIEW

File: pkgs/development/compilers/crystal/default.nix [hunk 1]
```diff
@@ -91,39 +93,44 @@ let
     ];
 
     FLAGS = [
-      "--release"
+      # "--release"
```
Comment (fpletz) [inline pkgs/development/compilers/crystal/default.nix]: Is that change intentional? If so, why? :)

Comment (manveru) [reply pkgs/development/compilers/crystal/default.nix]: Good catch :)

Comment (manveru) [reply pkgs/development/compilers/crystal/default.nix]: Had that there to speed up compiles during testing.

File: pkgs/development/compilers/crystal/default.nix [hunk 2]
```diff
@@ -56,6 +56,8 @@ let
       inherit sha256;
     };
 
+    outputs = [ "out" "lib" "bin" "share" ];
```
Comment (matthewbauer) [inline pkgs/development/compilers/crystal/default.nix]: "share" is a little bit unusual. For instance, we don't have a lib.getShare, like we do for lib.getDev, lib.getLib, and lib.getBin. Unless the share & out outputs are intentionally separate, I would recommend just putting share in $out.

Comment (matthewbauer) [reply pkgs/development/compilers/crystal/default.nix]: "doc" would be a little bit better if you would like.

Comment (manveru) [reply pkgs/development/compilers/crystal/default.nix]: `share` has stuff like shell completions, the manpage, a bunch of html docs, and the license... I'm not sure it matters since this gets all symlinked into `$out` anyway, but if you feel strongly about it i don't mind removing it either.

File: pkgs/development/compilers/crystal/default.nix [hunk 3]
```diff
@@ -97,33 +99,38 @@ let
...
+    CRYSTAL_LIBRARY_PATH = "${placeholder "lib"}/crystal";
 
     # We *have* to add `which` to the PATH or crystal is unable to build stuff
     # later if which is not available.
     installPhase = ''
       runHook preInstall
 
-      install -Dm755 .build/crystal $out/bin/crystal
-      wrapProgram $out/bin/crystal \
+      install -Dm755 .build/crystal $bin/bin/crystal
```
Comment (bew) [inline pkgs/development/compilers/crystal/default.nix]: Isn't it?
```suggestion
      install -Dm755 .build/crystal $bin/crystal
```
(ditto below)

**Review by matthewbauer — APPROVED** [review]
